src/pydvl/influence/torch/base.py

```diff
@@ -0,0 +1,710 @@
+
+    def jacobian_prod(
+        self,
+        batch: TorchBatch,
+        g: torch.Tensor,
+    ) -> torch.Tensor:
+        r"""
+        Computes the matrix-Jacobian product for the provided batch and input tensor.
+        Given the example in the class docstring, this means
+
+        $$ (\nabla_{\omega_{1}}\ell(\omega_1, \omega_2,
+            \text{batch.x}, \text{batch.y}),
+            \nabla_{\omega_{2}}\ell(\omega_1, \omega_2,
+            \text{batch.x}, \text{batch.y})) \cdot g^T$$
+
+        where g must be a tensor of shape $(K, d_1+d_2)$, so the resulting tensor
+        is of shape $(N, K)$.
+
+        Args:
+            batch: The batch of data for which to compute the Jacobian.
+            g: The tensor to be used in the matrix-Jacobian product
+                calculation.
+
+        Returns:
+            The resulting tensor from the matrix-Jacobian product computation.
+        """
+        result = self._jacobian_prod(batch.to(self.device), g.to(self.device))
+        if result.requires_grad:
+            result = result.detach()
+        return result
+
+    def flat_grads(self, batch: TorchBatch) -> torch.Tensor:
+        return flatten_dimensions(
+            self.grads(batch).values(), shape=(batch.x.shape[0], -1)
+        )
+
+    def flat_mixed_grads(self, batch: TorchBatch) -> torch.Tensor:
+        shape = (*batch.x.shape, -1)
+        return flatten_dimensions(self.mixed_grads(batch).values(), shape=shape)
+
+
+class OperatorBilinearForm(
+    BilinearForm[torch.Tensor, TorchBatch, TorchGradientProvider]
+):
+    r"""
+    Base class for bilinear forms based on an instance of
+    [TorchOperator][pydvl.influence.torch.operator.base.TorchOperator]. This means it
+    computes weighted inner products of the form:
+
+    $$ \langle \operatorname{Op}(x), y \rangle $$
+
+    """
+
+    def __init__(
+        self,
+        operator: "TensorOperator",
+    ):
+        self.operator = operator
+
+    def inner_prod(
+        self, left: torch.Tensor, right: Optional[torch.Tensor]
+    ) -> torch.Tensor:
+        r"""
+        Computes the weighted inner product of two vectors, i.e.
+
+        $$ \langle \operatorname{Op}(\text{left}), \text{right} \rangle $$
+
+        Args:
+            left: The first tensor in the inner product computation.
+            right: The second tensor, optional; if not provided, the inner product will
+                use `left` tensor for both arguments.
+
+        Returns:
+            A tensor representing the inner product.
+        """
+        if right is None:
+            right = left
+        if left.shape[0] <= right.shape[0]:
+            return self._inner_product(left, right)
+        return self._inner_product(right, left).T
+
+    def _inner_product(self, left: torch.Tensor, right: torch.Tensor) -> torch.Tensor:
+        left_result = self.operator.apply(left)
+
+        if left_result.ndim == right.ndim and left.shape[-1] == right.shape[-1]:
```

Comment:
I would prefer an explicit `else` over a second return statement. I find it most readable if methods have only one return statement.